# Gene drives

#### Creative Homework Assignment:

What features would you want to see in an online discussion platform devoted to guiding the development of gene drives? Assume that the researchers involved in the project are interested in soliciting public feedback before and during experiments so that they can better identify problems and redesign the technology. Please give specific examples of already-existing elements – if you want a discussion forum, should it be more like reddit, Quora, or something else? What should moderation be like?

What would it take to get you to regularly participate in such a community?

#### Design Homework Assignment:

Identify a problem that could be addressed using a CRISPR gene drive.

Which organism would you target and how would you alter it?

Why is a gene drive a good solution relative to other options?

What could go wrong? Don't go into detail, but list several possibilities.

Who should be involved in the discussion of whether to consider this application?

Design a basic but evolutionarily stable gene drive that should function in your organism.

(If you find this challenging, see the proposal to build a gene drive capable of eradicating schistosomiasis.

Your goal is to design a proof-of-principle experiment that will determine whether CRISPR gene drives can function efficiently in the target organism. Your drive system should not cause population suppression, carry any 'cargo' genes, or change the sequence of any protein produced by the organism. It should only spread itself.

If the genome sequence for your organism isn't available or is hard to work with in any way, use the nematode worm Caenorhabditis elegans instead – we may use your design for a hands-on experiment in next year's class.

1) Identify a gene that is important for fitness in your target organism. Why must gene drives target genes important for fitness? A literature search may be required. You can use NCBI for information on genes if you wish, though Wormbase is generally superior and easier to use.

2) Find CRISPR target sites in the 3' end of the gene using an online tool. For C. elegans and many other organisms, a good user-friendly tool is GT-Scan. Recommended parameters: set the high-specificity mismatch limit to 3 and leave all other settings at default values.

More exotic species require less user-friendly software. For bonus points, use sgRNAcas9, which allows you to analyze any downloaded genome sequence in fasta format. Warning: it's not particularly user-friendly.

List the relevant criteria for inclusion of each target sequence (e.g. potential off-targets etc) as provided by the program you used.

3) Would it be possible for you to safely build this gene drive in your current laboratory? Which confinement strategies and safeguards would you use? See Akbari et al. and especially here for recommended confinement strategies.

Extra credit:

4) Generate a version of the target gene that will be carried with your gene drive. Recode the target gene so that it encodes the same protein, but does not have the sequences you are targeting. Should you change only those codons, most, or all of them?
